Backend/learner/resnet18_vgg16.py

```diff
@@ -0,0 +1,156 @@
+        train_loader = DataLoader(train_dataset, batch_size=self.batch_size, shuffle=True)
+        test_loader = DataLoader(test_dataset, batch_size=self.batch_size, shuffle=False)
+        return train_loader, test_loader
+
+    def train(self):
+        for epoch in range(self.epochs):
+            running_loss = 0.0
+            for data in tqdm(self.train_loader, desc=f"Epoch {epoch+1}/{self.epochs}"):
+                inputs, labels = data
+                inputs, labels = inputs.to(self.device), labels.to(self.device)
+                self.optimizer.zero_grad()
+                outputs = self.model(inputs)
+                loss = self.criterion(outputs, labels)
+                loss.backward()
+                self.optimizer.step()
+                running_loss += loss.item()
+            epoch_loss = running_loss / len(self.train_loader)
+            self.epoch_losses.append(epoch_loss)
+            print(f"Epoch {epoch + 1}, Loss: {epoch_loss}")
+            self.evaluate(epoch)
+
+    def evaluate(self, epoch):
+        correct = 0
+        total = 0
+        with torch.no_grad():
+            for data in self.test_loader:
+                images, labels = data
+                images, labels = images.to(self.device), labels.to(self.device)
+                outputs = self.model(images)
+                _, predicted = torch.max(outputs.data, 1)
+                total += labels.size(0)
+                correct += (predicted == labels).sum().item()
+        accuracy = 100 * correct / total
+        self.epoch_accuracies.append(accuracy)
+        print(f'Accuracy after epoch {epoch + 1}: {accuracy}%')
+
+    def save_model(self, folder, model_name):
+        if not os.path.exists(folder):
+            os.makedirs(folder)
+        torch.save(self.model.state_dict(), os.path.join(folder, model_name))
+        print('Model saved successfully.')
+
+    def load_model(self, model_path):
+        self.model.load_state_dict(torch.load(model_path))
+        self.model.to(self.device)
+        self.model.eval()
+        print('Model loaded successfully.')
+
+    def plot_metrics(self):
+        plt.figure(figsize=(12, 5))
+        plt.subplot(1, 2, 1)
+        plt.plot(range(1, self.epochs + 1), self.epoch_losses, label='Training Loss')
+        plt.xlabel('Epoch')
+        plt.ylabel('Loss')
+        plt.title('Training Loss')
+
+        plt.subplot(1, 2, 2)
+        plt.plot(range(1, self.epochs + 1), self.epoch_accuracies, label='Accuracy', color='red')
+        plt.xlabel('Epoch')
+        plt.ylabel('Accuracy')
+        plt.title('Model Accuracy')
+        plt.show()
+
+    def predict(self, image_path):
+        image = Image.open(image_path).convert('RGB')  
+        image = self.transform(image).unsqueeze(0).to(self.device)
+        self.model.eval()
+        with torch.no_grad():
+            outputs = self.model(image)
+            _, predicted = torch.max(outputs.data, 1)
+        return self.dataset.classes[predicted.item()]
```

Comment:
Alright we are going to have merge conflict.

Reply:
why?

Reply:
We have merge conflicts when two branches works on same area 😥

Reply:
fixed this btw!